{
  "name": "retrying",
  "description": "Retrying\n=========================\n.. image:: https://travis-ci.org/rholder/retrying.png?branch=master\n    :target: https://travis-ci.org/rholder/retrying\n\n.. image:: https://badge.fury.io/py/retrying.png\n    :target: https://pypi.python.org/pypi/retrying\n\n.. image:: https://pypip.in/d/retrying/badge.png\n    :target: https://pypi.python.org/pypi/retrying\n\nRetrying is an Apache 2.0 licensed general-purpose retrying library, written in\nPython, to simplify the task of adding retry behavior to just about anything.\n\n\nThe simplest use case is retrying a flaky function whenever an Exception occurs\nuntil a value is returned.\n\n.. code-block:: python\n\n    import random\n    from retrying import retry\n\n    @retry\n    def do_something_unreliable():\n        if random.randint(0, 10) > 1:\n            raise IOError(\"Broken sauce, everything is hosed!!!111one\")\n        else:\n            return \"Awesome sauce!\"\n\n    print do_something_unreliable()\n\n\nFeatures\n--------\n\n- Generic Decorator API\n- Specify stop condition (i.e. limit by number of attempts)\n- Specify wait condition (i.e. exponential backoff sleeping between attempts)\n- Customize retrying on Exceptions\n- Customize retrying on expected returned result\n\n\nInstallation\n------------\n\nTo install retrying, simply:\n\n.. code-block:: bash\n\n    $ pip install retrying\n\nOr, if you absolutely must:\n\n.. code-block:: bash\n\n    $ easy_install retrying\n\nBut, you might regret that later.\n\n\nExamples\n----------\n\nAs you saw above, the default behavior is to retry forever without waiting.\n\n.. code-block:: python\n\n    @retry\n    def never_give_up_never_surrender():\n        print \"Retry forever ignoring Exceptions, don't wait between retries\"\n\n\nLet's be a little less persistent and set some boundaries, such as the number of attempts before giving up.\n\n.. code-block:: python\n\n    @retry(stop_max_attempt_number=7)\n    def stop_after_7_attempts():\n        print \"Stopping after 7 attempts\"\n\nWe don't have all day, so let's set a boundary for how long we should be retrying stuff.\n\n.. code-block:: python\n\n    @retry(stop_max_delay=10000)\n    def stop_after_10_s():\n        print \"Stopping after 10 seconds\"\n\nMost things don't like to be polled as fast as possible, so let's just wait 2 seconds between retries.\n\n.. code-block:: python\n\n    @retry(wait_fixed=2000)\n    def wait_2_s():\n        print \"Wait 2 second between retries\"\n\n\nSome things perform best with a bit of randomness injected.\n\n.. code-block:: python\n\n    @retry(wait_random_min=1000, wait_random_max=2000)\n    def wait_random_1_to_2_s():\n        print \"Randomly wait 1 to 2 seconds between retries\"\n\nThen again, it's hard to beat exponential backoff when retrying distributed services and other remote endpoints.\n\n.. code-block:: python\n\n    @retry(wait_exponential_multiplier=1000, wait_exponential_max=10000)\n    def wait_exponential_1000():\n        print \"Wait 2^x * 1000 milliseconds between each retry, up to 10 seconds, then 10 seconds afterwards\"\n\n\nWe have a few options for dealing with retries that raise specific or general exceptions, as in the cases here.\n\n.. code-block:: python\n\n    def retry_if_io_error(exception):\n        \"\"\"Return True if we should retry (in this case when it's an IOError), False otherwise\"\"\"\n        return isinstance(exception, IOError)\n\n    @retry(retry_on_exception=retry_if_io_error)\n    def might_io_error():\n        print \"Retry forever with no wait if an IOError occurs, raise any other errors\"\n\n    @retry(retry_on_exception=retry_if_io_error, wrap_exception=True)\n    def only_raise_retry_error_when_not_io_error():\n        print \"Retry forever with no wait if an IOError occurs, raise any other errors wrapped in RetryError\"\n\nWe can also use the result of the function to alter the behavior of retrying.\n\n.. code-block:: python\n\n    def retry_if_result_none(result):\n        \"\"\"Return True if we should retry (in this case when result is None), False otherwise\"\"\"\n        return result is None\n\n    @retry(retry_on_result=retry_if_result_none)\n    def might_return_none():\n        print \"Retry forever ignoring Exceptions with no wait if return value is None\"\n\n\nAny combination of stop, wait, etc. is also supported to give you the freedom to mix and match.\n\nContribute\n----------\n\n#. Check for open issues or open a fresh issue to start a discussion around a feature idea or a bug.\n#. Fork `the repository`_ on GitHub to start making your changes to the **master** branch (or branch off of it).\n#. Write a test which shows that the bug was fixed or that the feature works as expected.\n#. Send a pull request and bug the maintainer until it gets merged and published. :) Make sure to add yourself to AUTHORS_.\n\n.. _`the repository`: http://github.com/rholder/retrying\n.. _AUTHORS: https://github.com/rholder/retrying/blob/master/AUTHORS.rst\n\n\n.. :changelog:\n\nHistory\n-------\n1.3.3 (2014-12-14)\n++++++++++++++++++\n- Add minimum six version of 1.7.0 since anything less will break things\n\n1.3.2 (2014-11-09)\n++++++++++++++++++\n- Ensure we wrap the decorated functions to prevent information loss\n- Allow a jitter value to be passed in\n\n1.3.1 (2014-09-30)\n++++++++++++++++++\n- Add requirements.txt to MANIFEST.in to fix pip installs\n\n1.3.0 (2014-09-30)\n++++++++++++++++++\n- Add upstream six dependency, remove embedded six functionality\n\n1.2.3 (2014-08-25)\n++++++++++++++++++\n- Add support for custom wait and stop functions\n\n1.2.2 (2014-06-20)\n++++++++++++++++++\n- Bug fix to not raise a RetryError on failure when exceptions aren't being wrapped\n\n1.2.1 (2014-05-05)\n++++++++++++++++++\n- Bug fix for explicitly passing in a wait type\n\n1.2.0 (2014-05-04)\n++++++++++++++++++\n- Remove the need for explicit specification of stop/wait types when they can be inferred\n- Add a little checking for exception propagation\n\n1.1.0 (2014-03-31)\n++++++++++++++++++\n- Added proper exception propagation through reraising with Python 2.6, 2.7, and 3.2 compatibility\n- Update test suite for behavior changes\n\n1.0.1 (2013-03-20)\n++++++++++++++++++\n- Fixed a bug where classes not extending from the Python exception hierarchy could slip through\n- Update test suite for custom Python exceptions\n\n1.0.0 (2013-01-21)\n++++++++++++++++++\n- First stable, tested version now exists\n- Apache 2.0 license applied\n- Sanitizing some setup.py and test suite running\n- Added Travis CI support",
  "requires_dist": null,
  "requires_python": null,
  "current_version": "1.3.3",
  "released_versions": [
    "1.0.0",
    "1.0.1",
    "1.1.0",
    "1.2.0",
    "1.2.1",
    "1.2.2",
    "1.2.3",
    "1.3.0",
    "1.3.1",
    "1.3.2",
    "1.3.3"
  ]
}